stan/math/prim/meta/is_eigen.hpp

```diff
@@ -2,84 +2,40 @@
 #define STAN_MATH_PRIM_META_IS_EIGEN_HPP
 
 #include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta/bool_constant.hpp>
 #include <stan/math/prim/meta/disjunction.hpp>
 #include <type_traits>
 
 namespace stan {
 
-/** \ingroup type_trait
- * Base implementation to check whether a type is derived from EigenBase
+/** \addtogroup type_trait
+ *  @{
  */
-template <typename T, typename = void>
-struct is_eigen : std::false_type {};
 
-namespace internal {
 /**
- * Underlying implementation to check if a type is derived from EigenBase
- */
-template <typename T>
-struct is_eigen_base
-    : std::integral_constant<bool,
-                             std::is_base_of<Eigen::EigenBase<T>, T>::value> {};
-
-template <typename T>
-struct is_eigen_base<Eigen::DenseBase<T>> : std::true_type {};
-
-template <typename T>
-struct is_eigen_base<Eigen::MatrixBase<T>> : std::true_type {};
-
-template <typename T>
-struct is_eigen_base<Eigen::EigenBase<T>> : std::true_type {};
-
-}  // namespace internal
+ * Check if type derives from `EigenBase`
+ * @tparam T Type to check if it is derived from `EigenBase`
+ * @tparam Enable used for SFINAE deduction.
+ **/
+template <typename T, typename Enable = void>
+struct is_eigen : std::false_type {};
 
-/**
- * Checks whether type T is derived from EigenBase. If true this will have a
- * static member function named value with a type of true, else value is false.
- */
 template <typename T>
-struct is_eigen<
-    T, std::enable_if_t<internal::is_eigen_base<std::decay_t<T>>::value>>
+struct is_eigen<T, std::enable_if_t<std::is_base_of<
+                       Eigen::EigenBase<typename std::decay_t<T>::PlainObject>,
+                       typename std::decay_t<T>::PlainObject>::value>>
```

Comment:
Just a general comment, but it might make these templates more readable if the ```typename std::decay_t<T>::PlainObject``` is moved up to the template definition:
```
template <typename T, typename PlainType = typename std::decay_t<T>::PlainObject>
struct is_eigen<T, std::enable_if_t<std::is_base_of<
                       Eigen::EigenBase<PlainType>,
                       PlainType>::value>>
```

Reply:
Agree!!

Reply:
Actually idt this would work since it would set the default of the template value but I have to grab `PlainObject` in one specialization and `MatrixType` in another